envoy-control-tests/src/main/resources/lua_spec/ingress_handler_spec.lua

```diff
@@ -0,0 +1,227 @@
+            ssl = function() return https or nil end
+        } end,
+        logInfo = logInfoMock
+    }
+end
+
+describe("envoy_on_request:", function()
+    it("should set dynamic metadata", function()
+        -- given
+        local headers = {
+            [':path'] = '/path',
+            [':method'] = 'GET',
+            ['x-service-name'] = 'lorem-service',
+            ['x-forwarded-for'] = "127.0.4.3"
+        }
+
+        local handle = handlerMock(headers)
+        local metadata = handle:streamInfo():dynamicMetadata()
+
+        -- when
+        envoy_on_request(handle)
+
+        -- then
+        assert.spy(metadata.set).was_called_with(_, "envoy.filters.http.lua", "request.info.path", "/path")
+        assert.spy(metadata.set).was_called_with(_, "envoy.filters.http.lua", "request.info.method", "GET")
+        assert.spy(metadata.set).was_called_with(_, "envoy.filters.http.lua", "request.info.service_name", "lorem-service")
+        assert.spy(metadata.set).was_called_with(_, "envoy.filters.http.lua", "request.info.xff_header", "127.0.4.3")
+    end)
+end)
+
+describe("envoy_on_response:", function()
+    local headers
+    local metadata
+    local ssl
+
+    before_each(function ()
+        headers = {
+            [':status'] = '403'
+        }
+        metadata = {
+            ['envoy.filters.http.rbac'] = {
+                ['shadow_engine_result'] = 'denied'
+            },
+            ['envoy.filters.http.lua'] = {
+                ['request.info.service_name'] = 'service-first',
+                ['request.info.path'] = '/path?query=val',
+                ['request.info.method'] = 'POST',
+                ['request.info.xff_header'] = '127.1.1.3',
+
+            }
+        }
+        ssl = true
+    end)
+
+    describe("should log unauthorized requests:", function ()
+
+        it("https request", function ()
+            -- given
+            local handle = handlerMock(headers, metadata, ssl)
+
+            -- when
+            envoy_on_response(handle)
+
+            -- then
+            assert.spy(handle.logInfo).was_called_with(_,
+              "\nUnauthorized request: method = POST, path = '/path?query=val', "..
+              "clientIp = 127.1.1.3, clientName = service-first, protocol = https, statusCode = 403")
+            assert.spy(handle.logInfo).was_called(1)
+        end)
+
+        it("http request", function ()
+            -- given
+            ssl = false
+            local handle = handlerMock(headers, metadata, ssl)
+
+            -- when
+            envoy_on_response(handle)
+
+            -- then
+            assert.spy(handle.logInfo).was_called_with(_,
+              "\nUnauthorized request: method = POST, path = '/path?query=val', "..
+                "clientIp = 127.1.1.3, clientName = service-first, protocol = http, statusCode = 403")
+            assert.spy(handle.logInfo).was_called(1)
+        end)
+
+        it("allowed & logged request", function ()
```

Comment:
How does this test differ from `https request`?

Reply:
These tests specify business requirements. One of the requirements is to log request which is unauthorized, but accepted. It turns out, that in implementation it only differs by `:status` header from request that was rejected. So it doesn't make much sense in terms of testing, but it make sense as a business requirement specification and documentation IMO, so I would keep it here.